raphtory/src/db/graph/graph.rs

```diff
@@ -726,6 +729,337 @@ mod db_tests {
+
+        let gg = Graph::new();
+        gg.add_node(1, "Q", NO_PROPS, None).unwrap();
+        let res = gg.import_nodes_as(vec![&g_a, &g_b], vec!["P", "Q"], false);
+        match res {
+            Err(NodeExistsError(id)) => {
+                assert_eq!(
+                    id.to_string(),
+                    "Q",
+                    "Unexpected node ID for NodeExistsError"
+                );
+            }
+            Err(e) => panic!("Unexpected error: {:?}", e),
+            Ok(_) => panic!("Expected error but got Ok"),
+        }
+        let mut nodes = gg.nodes().name().collect_vec();
+        nodes.sort();
+        assert_eq!(nodes, vec!["P", "Q"]); // Nodes up until first failure are imported
+        let y = gg.node("Q").unwrap();
+        assert_eq!(y.name(), "Q");
+        assert_eq!(y.history(), vec![1]);
+        assert_eq!(y.properties().get("temp"), None);
+        assert_eq!(y.properties().constant().get("con"), None);
+    }
+
+    #[test]
+    fn import_nodes_as_force() {
+        let g = Graph::new();
+        let g_a = g.add_node(0, "A", NO_PROPS, None).unwrap();
+        let g_b = g
+            .add_node(1, "B", vec![("temp".to_string(), Prop::Bool(true))], None)
+            .unwrap();
+        let _ = g_b.add_constant_properties(vec![("con".to_string(), Prop::I64(11))]);
+
+        let gg = Graph::new();
+        gg.add_node(1, "Q", NO_PROPS, None).unwrap();
+        let _ = gg
+            .import_nodes_as(vec![&g_a, &g_b], vec!["P", "Q"], true)
+            .unwrap();
+        let mut nodes = gg.nodes().name().collect_vec();
+        nodes.sort();
+        assert_eq!(nodes, vec!["P", "Q"]);
+        let y = gg.node("Q").unwrap();
+        assert_eq!(y.name(), "Q");
+        assert_eq!(y.history(), vec![1]);
+        assert_eq!(y.properties().get("temp").unwrap(), Prop::Bool(true));
+        assert_eq!(y.properties().constant().get("con").unwrap(), Prop::I64(11));
+    }
+
+    #[test]
+    fn import_edge_as() {
+        let g = Graph::new();
+        let g_a = g.add_node(0, "A", NO_PROPS, None).unwrap();
+        let g_b = g
+            .add_node(1, "B", vec![("temp".to_string(), Prop::Bool(true))], None)
+            .unwrap();
+        let _ = g_b.add_constant_properties(vec![("con".to_string(), Prop::I64(11))]);
+        let e_a_b = g
+            .add_edge(
+                2,
+                "A",
+                "B",
+                vec![("e_temp".to_string(), Prop::Bool(false))],
+                None,
+            )
+            .unwrap();
+
+        let gg = Graph::new();
+        let e = gg.add_edge(1, "X", "Y", NO_PROPS, None).unwrap();
+        let res = gg.import_edge_as(&e_a_b, ("X", "Y"), false);
+        match res {
+            Err(EdgeExistsError(src_id, dst_id)) => {
+                assert_eq!(
+                    src_id.to_string(),
+                    "X",
+                    "Unexpected node ID for NodeExistsError"
```

Comment:
This reads like there is a problem becuase the nodes existed, not because the edge was there no?